pkg/workspace/estimator/advancednodesestimator/estimator.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) KAITO authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package advancednodesestimator
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	kaitov1beta1 "github.com/kaito-project/kaito/api/v1beta1"
+	"github.com/kaito-project/kaito/pkg/utils"
+	"github.com/kaito-project/kaito/pkg/utils/consts"
+	"github.com/kaito-project/kaito/pkg/utils/plugin"
+)
+
+// AdvancedNodesEstimator estimates node count based on SKU memory and model memory requirement
+type AdvancedNodesEstimator struct {
+	// no fields needed
+}
+
+func (c *AdvancedNodesEstimator) Name() string {
+	return "advanced"
+}
+
+func (c *AdvancedNodesEstimator) EstimateNodeCount(ctx context.Context, workspace *kaitov1beta1.Workspace) (int32, error) {
+	// If inference is not configured, default to resource count or 1
+	if workspace.Inference == nil || workspace.Inference.Preset == nil || workspace.Inference.Preset.Name == "" {
+		//nolint:staticcheck //SA1019: deprecate Resource.Count field
+		if workspace.Resource.Count != nil {
+			//nolint:staticcheck //SA1019: deprecate Resource.Count field
+			return int32(*workspace.Resource.Count), nil
+		}
+		return 1, nil
+	}
+
+	presetName := string(workspace.Inference.Preset.Name)
+	model := plugin.KaitoModelRegister.MustGet(presetName)
+
+	gpuConfig, err := utils.GetGPUConfigBySKU(workspace.Resource.InstanceType)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get GPU config for instance type %s: %w", workspace.Resource.InstanceType, err)
+	}
+	if gpuConfig == nil {
+		return 0, fmt.Errorf("GPU config is nil for instance type %s", workspace.Resource.InstanceType)
+	}
+
+	// Start with the user-requested node count (default is 1)
+	//nolint:staticcheck //SA1019: deprecate Resource.Count field
+	nodeCountPerReplica := 1 // Default to 1 if not specified
+	//nolint:staticcheck //SA1019: deprecate Resource.Count field
+	if workspace.Resource.Count != nil {
+		//nolint:staticcheck //SA1019: deprecate Resource.Count field
+		nodeCountPerReplica = int(*workspace.Resource.Count)
+	}
+
+	// Use default max-model-len value
+	maxModelLen := 2048 // Default value
```

Comment:
Not every model has 2048 max-model-len. How do we account for longer model length?

Reply:
@chewong For now, ever model we support is more than 2048, we just set it as 2048 if user does not have an input to calculate the node